contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,302 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v5.0.0) (access/AccessControl.sol)
+
+pragma solidity ^0.8.20;
+
+import {Context} from "@openzeppelin/contracts/utils/Context.sol";
+import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+
+/**
+ * @dev An enhanced version of OpenZeppelin's AccessControl that allows for:
+ * 
+ * - Resource-based roles.
+ * - Root resource override (0x0) - role assignments in the root resource auto-apply to all resources.
+ * - Max 256 roles - stored as a bitmap in uint256.
+ */
+abstract contract EnhancedAccessControl is Context, ERC165 {
+    error EnhancedAccessControlUnauthorizedAccountRole(bytes32 resource, uint256 role, address account);
+    error EnhancedAccessControlUnauthorizedAccountAdminRole(bytes32 resource, uint256 role, address account);
+    error EnhancedAccessControlBadConfirmation();
+    error EnhancedAccessControlLockedRole(bytes32 resource, uint256 role);
+
+    event EnhancedAccessControlRoleAdminChanged(uint256 role, uint256 previousAdminRole, uint256 newAdminRole);
+    event EnhancedAccessControlRoleGranted(bytes32 resource, uint256 role, address account, address sender);
+    event EnhancedAccessControlRolesGranted(bytes32 resource, uint256 roleBitmap, address account, address sender);
+    event EnhancedAccessControlRolesCopied(bytes32 srcResource, bytes32 dstResource, address account, address sender, uint256 roleBitmap);
+    event EnhancedAccessControlRolesRevoked(bytes32 resource, uint256 role, address account, address sender);
+    event EnhancedAccessControlAllRolesRevoked(bytes32 resource, address account, address sender);
+    event EnhancedAccessControlRolesLocked(bytes32 resource, uint256 roleBitmap);
+
+    /** 
+     * @dev admin role that controls a given role. 
+     * RoleId -> AdminRoleId
+     */
+    mapping(uint256 role => uint256 adminRole) private _adminRoles;
+
+    /**
+     * @dev user roles within a resource stored as a bitmap.
+     * Resource -> User -> RoleBitmap
+     */
+    mapping(bytes32 resource => mapping(address account => uint256 roleBitmap)) private _roles;
+
+    /**
+     * @dev locked roles within a resource stored as a bitmap.
+     * Resource -> LockedRoleBitmap
+     */
+    mapping(bytes32 resource => uint256 lockedRoleBitmap) private _lockedRoles;
```

Comment:
Is this necessary? If a role for a given resource has no admins, it's locked implicitly.

Reply:
The issue is that the `DEFAULT_ADMIN_ROLE` can always revoke a role, thus I figured we needed an explicit locking mechanism to prevent that.

Reply:
With the admin-bit-for-each-role I suggested above, there is no DEFAULT_ADMIN_ROLE - only individual admins for each role.